### ISSUE TITLE
Duplicate sky & target id in mock ids.  

### ISSUE THREAD

**Issue by michaelJwilson** [desihub/desitarget#609]
This is first off a question, but potentially a bug. 

I've been generating mock targets & mock skies successfully for the 1% survey for a while.  The latest incarnation and new feature includes two disjoint fields, one in the NGC & SGC.  But I'm running into a problem with this one.

Fiberassign is throwing an error at duplicate TARGETIDS that hasn't occured on a single footprint run.   I've checked the dark and sky files and each id appears only once.  But the id picked on by fiberassign is common to both. 

The relevant line is 1046 under mock/build.py where 

```
objid = np.arange(nobj + nsky)

targetid = encode_targetid(objid=objid, brickid=healpix, mock=1)
```

called for each healpixel.  I'll rewrite the sky targetids and dig into the encoding, but if the answer is obvious to anyone then let me know.  Thanks.

**Comment by geordie666** [comment]
I would imagine that the sky objects should simply have:
```
targetid = encode_targetid(objid=objid, brickid=healpix, mock=1, sky=1)
```
at least, `sky=1` is how I guarantee unique sky `TARGETID`s for the "real" data.

**Comment by michaelJwilson** [comment]
geI followed the logic of the original script, e.g. objid = arange(nsky) when split by healpixel, but including sky=1 when overwriting the targetids in the sky.fits I had previously generated and it worked a treat for fiberassign etc.  Thanks!  I'll upgrade this to a bug and include the fix in the maskbits PR.    

The remaining mystery is why this problem hasn't shown up in the full footprint runs that have happened, but I'll assume that's irrelevant.

**Comment by moustakas** [comment]
I would like to look, too, please. I have a vague recollection that the
code is like that for a very specific reason. I also owe feedback on the
maskbits PR, sorry for the delay!

**Comment by michaelJwilson** [comment]
Ok, I'd propose the simple change:

```
allid = np.arange(nobj + nsky)
obj_targetid = encode_targetid(objid=allid[:nobj], brickid=healpix, mock=1)
sky_targetid = encode_targetid(objid=allid[nobj:], brickid=healpix, mock=1, sky=1)
targetid = np.concatenate([obj_targetid, sky_targetid])
```
But hold off until maskbits is closed as a PR andopen this as a new one if necessary - once John gets to jog his memory.  Potentially you might be remembering the comment on this line:

# Assign TARGETID using the healpixel number, not BRICKID, otherwise we'll
# end up with duplicate TARGETID values.

I'll patch it in the meantime.  Thanks.

**Comment by moustakas** [comment]
Maybe I need to look at the guts of `encode_targetid`---and I understand that @michaelJwilson has found evidence to the contrary---but I don't understand why 
```
objid = np.arange(nobj + nsky)
targetid = encode_targetid(objid=objid, brickid=healpix, mock=1)
```
doesn't guarantee unique targetIDs over the full footprint, since `healpix`s (like `brickid`s) are unique over the full sky and `objid`, within a given healpixel, is guaranteed to not be duplicated.

What am I not understanding? (Just about this issue--be kind!)

**Comment by geordie666** [comment]
@michaelJwilson will have to show you his specific use-case to answer this.

I think you understand how `encode_targetid` works, and I'm not even just being kind.

If there is a lack of uniqueness, it would (I think) be something to do with how you're splitting the generation of skies and objects (i.e. nsky=0, nobj=N when you generate objects but then nobj=0, nsky=N when you generate skies). Or, perhaps you could be exceeding the allowed number of bits in `objid` (I think that would flag an error though).

**Comment by michaelJwilson** [comment]
Apologies in advance if I screwed this up, I agree it's weird. 

I had a concern that I used to copy skies files across from the data to have locally. 
But I've checked the timestamps N>1 times and the skies was created at the same
time as the targets (join).  

They're here:
``` /global/homes/m/mjwilson/desi/survey-validation/svdc-spring2020g-onepercent/run/targets ```

in case that's useful.

I was having some problems with the mpi runs so I've been running serially so it's a
bit of a pain for me to reproduce this to confirm.  Obviously, we can correct that.
